Add support for actions on hover

Astrono2 opened the pull request:
<!-- Please read our contributing guide before opening a PR: https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md -->

## What type of PR is this? (check all applicable)

* [ ] Refactor
* [x] Feature
* [ ] Bug Fix
* [ ] Optimization
* [ ] Documentation Update
* [ ] Other: *Replace this with a description of the type of this PR*

## Description
<!--
  Document user-facing changes in this PR (for example: new config options, changed behavior).

  You can also motivate design decisions here.
-->
This will allow for actions to be triggered on hover. Hover actions will work like toggle actions, meaning that the module is responsible for keeping track of its state, the action call only notifies of a change in state.

## Related Issues & Documents
<!-- For example: Fixes #1234, Closes #6789 -->
Closes #1064

## Documentation (check all applicable)

* [x] This PR requires changes to the Wiki documentation (describe the changes)
* [x] This PR requires changes to the documentation inside the git repo (please add them to the PR).
* [ ] Does not require documentation changes

The feature is far for complete, so I can't predict the documentation changes. But there will certainly be some and I'll update this before marking it as ready for review.

patrick96 reviewed (commented):
The approach looks solid.

I am right now thinking about how users would use this. I think one of the main use-cases would be to spawn some sort of popup to display some information. Basically launching a program that has to stop running once the mouse leaves the hover region or the bar again.

This would mean we would need to detect and respond to the mouse leaving the hover region or bar and send some signal to the process. This could get quite complicated.
What do you think of this?

What kind of use-cases did you have in mind?

Astrono2 commented:
> The approach looks solid.
> 
> I am right now thinking about how users would use this. I think one of the main use-cases would be to spawn some sort of popup to display some information. Basically launching a program that has to stop running once the mouse leaves the hover region or the bar again.
> 
> This would mean we would need to detect and respond to the mouse leaving the hover region or bar and send some signal to the process. This could get quite complicated.
> What do you think of this?
> 
> What kind of use-cases did you have in mind?

This already detects the mouse leaving the hover region. I'm gonna add detecting leaving the bar next.
When the mouse leaves the hover region, it detects that it's hovering a new action (including no action). It then sends a hover signal to the previously hovered action. The only problem with this would be if two modules with the exact same action string for hover were next to each other.

I have in mind tooltips mostly. But also something like mapping #date.toggle to hover instead of click-left.

Astrono2 commented on src/components/bar.cpp:
```diff
@@ -656,8 +657,25 @@ void bar::handle(const evt::motion_notify& evt) {
 
   m_log.trace("bar: Detected motion: %i at pos(%i, %i)", evt->detail, evt->event_x, evt->event_y);
-#if WITH_XCURSOR
   m_motion_pos = evt->event_x;
```
Sorry, what can be a local variable? m_motion_pos?

patrick96 replied:
Exactly. I just saw that you moved it around and noticed that it doesn't need to be a class member.

Astrono2 replied:
Okay, I'll make it local

Astrono2 commented on src/components/bar.cpp:
```diff
@@ -656,8 +657,25 @@ void bar::handle(const evt::motion_notify& evt) {
+  if (!string_util::compare(m_hover_act, m_last_hover_act)) {
+    m_log.trace("bar: Switched hover action from '%s' to '%s'", m_last_hover_act, m_hover_act);
+    if(m_last_hover_act != "") {
+      m_sig.emit(button_press{m_last_hover_act});
```
Because hover works as a toggle. A hover action lets the module know that its hover state changed, it's up to the module to keep track of that state. I think that's the simplest way to handle it.

Astrono2 replied:
Speaking of, I'm having a big problem: you can't emit button press signals twice in the same method, the controller swallows the second one. I'm sure there's a simple fix, but I'm not very familiar with the codebase yet.
@patrick96 could you give me a hand? Is there a simple way to enqueue the second signal? Or a worse alternative, emit one and then stall until the next iteration of the eventqueue worker loop?
For reference, this isn't a problem when going from hovering a module to hovering nothing, because there is no second action to be called, but it is when having two modules with hover next to each other.

patrick96 replied:
> Because hover works as a toggle. A hover action lets the module know that its hover state changed, it's up to the module to keep track of that state. I think that's the simplest way to handle it.

I see. In that case, do you think it would make more sense to have separate "mouse enter" and "mouse leave" events?

> Speaking of, I'm having a big problem: you can't emit button press signals twice in the same method, the controller swallows the second one.

Yeah, the enqueueing of commands isn't great. Currently we store the command in `m_inputdata` and swallow all other commands as long as it has not been handled and the variable reset.
This could probably easily be replaced with a queue of strings where we place incoming commands at the end and `process_inputdata` takes one command (or all of them) from the front and processes them.

Astrono2 replied:
Yeah that's a good idea, but I think it's a bit outside of my scope.
I'll open an issue requesting that rework and put this on hold.

> I see. In that case, do you think it would make more sense to have separate "mouse enter" and "mouse leave" events?

I'm not sure. The problem with that would be that then the action would change. It can be taken into consideration, but I don't think the extra complexity it would need is worth it.

patrick96 replied:
> I'm not sure. The problem with that would be that then the action would change.

I don't think this would be very complex to solve. Right now we emit button presses when the current hover action changes. One for the new one and one for the old one. We would do the same thing but store both the current "mouse enter" action (so that we don't run on every mouse move, only the first time) and the current "mouse leave" action (so that we can detect when we leave a region and have to run the old leave action).

I'm just thinking about the usability. It's one thing if it's just the module itself that's keeping state, that's really easy. But as soon as the user wants to run something only while in the hover region, keeping state becomes very cumbersome.

Astrono2 commented on src/components/bar.cpp:
```diff
@@ -656,8 +657,25 @@ void bar::handle(const evt::motion_notify& evt) {
+    m_hover_act = m_action_ctxt->get_action(m_hover_act_idx);
+    m_log.trace("bar: Detected hover action %s", m_hover_act);
+  }
+  if (!string_util::compare(m_hover_act, m_last_hover_act)) {
```
Oh okay, I just saw that there was a `string_util::compare` and used it to keep my code similar to the existing one lol. I'll change that.

Astrono2 commented:
Oh and speaking of detecting when hovering a different action than before, I added two m_log.trace for debugging.
[`m_log.trace("bar: Detected hover action %s", m_hover_act);`](https://github.com/Astrono2/polybar/blob/c4d0a68521bc5ec0e137527bbd79c29a86722d6c/src/components/bar.cpp#L666) I think is too annoying to keep.
But [`m_log.trace("bar: Switched hover action from '%s' to '%s'", m_last_hover_act, m_hover_act);`](https://github.com/Astrono2/polybar/blob/c4d0a68521bc5ec0e137527bbd79c29a86722d6c/src/components/bar.cpp#L669) could be left in.

patrick96 commented:
> The only problem with this would be if two modules with the exact same action string for hover were next to each other.

I think we could then count this as one big hover region.

Astrono2 commented:
> I think we could then count this as one big hover region.

Luckily for me, that's the current behavior lol

patrick96 commented:
Just FYI, I won't be able to work on polybar for a few weeks.
